pkg/kedge/grpc/director/director.go

```diff
@@ -8,27 +8,56 @@ import (
 	"github.com/grpc-ecosystem/go-grpc-middleware/tags"
 	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
 	"github.com/improbable-eng/kedge/pkg/grpcutils"
+	"github.com/improbable-eng/kedge/pkg/kedge/common"
 	"github.com/improbable-eng/kedge/pkg/kedge/grpc/backendpool"
 	"github.com/improbable-eng/kedge/pkg/kedge/grpc/director/router"
 	"github.com/mwitkow/grpc-proxy/proxy"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
 // New builds a StreamDirector based off a backend pool and a router.
-func New(pool backendpool.Pool, router router.Router) proxy.StreamDirector {
+func New(pool backendpool.Pool, adhocRouter common.Addresser, grpcRouter router.Router) proxy.StreamDirector {
 	return func(ctx context.Context, fullMethodName string) (context.Context, *grpc.ClientConn, error) {
-		beName, err := router.Route(ctx, fullMethodName)
+		beName, err := grpcRouter.Route(ctx, fullMethodName)
+
+		// Try adhoc router if RouteNotFound
+		if err == router.ErrRouteNotFound {
+			ipPort, err := adhocRouter.Address(metautils.ExtractIncoming(ctx).Get(":authority"))
+			if err != nil {
+				return ctx, nil, errors.Wrap(err, "failed to proxy request using adhoc router")
```

Comment:
this is not always true message. Most likely we will get `router.ErrRouteNotFound` if no adhoc is specified. 

I would actually vote for not wrapping this error. It is already nicely wrapped in `Address` (`adhoc` prefix)